projects/table-builder/src/lib/directives/multi-sort.directive.ts

```diff
@@ -13,56 +14,33 @@ import { scan, filter, map } from 'rxjs/operators';
 })
 export class MultiSortDirective extends MatSort implements OnInit, OnDestroy {
   @Output() readonly multiSortChange: EventEmitter<Sort[]> = new EventEmitter<Sort[]>();
-
-  @Input() rules$: Observable<Sort[]> = new Subject();
-
-  @Input() set  reSort (val){
-    this.initRules();
-  } 
   rules: Sort[] = [];
-  preSortRules: Sort[] = [];
   private SubRef: Subscription;
 
+  constructor(private state: TableStateManager) {
+    super();
+  }
+
   ngOnInit() {
-    this.SubRef = this.rules$.pipe(
-      scan((acc,curr)=>({prev: acc.curr, curr }),{prev: null, curr: null} as {prev: Sort[], curr: Sort[]}),
-      filter(acc => !acc.prev || acc.curr.some(sort => !acc.prev.some(prevSort => sort.active === prevSort.active && sort.direction === prevSort.direction))),
-      map(acc => acc.curr)
+    this.SubRef = this.state.state$.pipe(
+      map( state => state.sorted ),
+      distinctUntilChanged(),
     ).subscribe( rules => {
-      this.preSortRules = rules;
-      this.initRules();
+      this.rules = rules;
+      this.multiSortChange.emit(this.rules);
     });
     super.ngOnInit();
   }
 
-  initRules(){
-    if (this.preSortRules.length ) {
-      this.active = null;
-      const initRules = [...this.preSortRules];
-      const firstRule = initRules.shift();
-      this.rules = initRules;
-      this.sort({id: firstRule.active, start: firstRule.direction || 'asc', disableClear: false});
-    }
-  }
 
   ngOnDestroy() {
     this.SubRef.unsubscribe();
     super.ngOnDestroy();
   }
 
   sort(sortable: MatSortable): void {
-
-    this.rules = this.rules.filter(r => r.active !== sortable.id);
-
-    if (this.active !== sortable.id) {
-      this.rules.unshift({ active: sortable.id, direction: sortable.start ? sortable.start : this.start });
-    } else {
-      const newDirection = this.getNextSortDirection(sortable);
-      if (newDirection) {
-        this.rules.unshift({ active: this.active, direction: newDirection });
-      }
-    }
+    const direction = this.active !== sortable.id ?  sortable.start ?? this.start : this.getNextSortDirection(sortable);
+    this.state.setSort(sortable.id,direction);
     super.sort(sortable);
-    this.multiSortChange.emit(this.rules);
```

Comment:
Why are we not emitting the change anymore?

Reply:
It was not in use.
You can always listen to the store instead of the directive which is the appropriate thing to do anyway.